File: BOJ/2157.cpp
#include<bits/stdc++.h>
#define f(i,l,r) for(int i=l;i<r;++i)
using namespace std;

const int INF=-3e6;
struct info{int u,c;};
vector<info> a[301];
int s[301][300];

int sol(int i,int j){
	if(j<0)return INF;
	if(i==1)return 0;
	int &ref=s[i][j];
	if(ref!=-1)return ref;
	
	ref=INF;
	for(info &in:a[i])ref=max(ref,sol(in.u,j-1)+in.c);
	return ref;
}

int main(){
	ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	int n,m,k;cin>>n>>m>>k;
	f(i,0,k){
		int u,v,c;cin>>u>>v>>c;
		if(u<v)a[v].push_back({u,c});
	}
	memset(s, -1, sizeof(s));
	cout<<sol(n,m-1);
	return 0;
}

File: BOJ/2157.swift
func sol(){
	let nmk=readLine()!.split(separator: " ").map{Int($0)!},n=nmk[0],m=nmk[1],k=nmk[2],INF = Int(-3e6)
	var a=Array(repeating: [(Int,Int)](), count: n+1),s=Array(repeating: Array(repeating: -1, count: m), count: n+1)
	for _ in 1...k{
		let info=readLine()!.split(separator: " ").map{Int($0)!}
		if info[1]>info[0] {a[info[1]].append((info[0],info[2]))}
	}
	
	func maxSum(_ i:Int,_ j:Int)->Int{
		if j<0{return INF}
		if i==1{return 0}
		if s[i][j] != -1{return s[i][j]}
		s[i][j]=INF
		for k in 0..<a[i].count{
			s[i][j]=max(s[i][j],maxSum(a[i][k].0, j-1)+a[i][k].1)
		}
		return s[i][j]
	}
	print(maxSum(n, m-1))
}
//sol()
